File: JavaScript/35_SearchInsertPosition.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var searchInsert = function(nums, target) {
    let l = 0;
    let r = nums.length - 1;
    let m;

    if (target > nums[r]) {
        return r + 1;
    } else if (target < nums[l]) {
        return l;
    }

    while (l <= r) {
        m = Math.floor((l + r) / 2);
        if (nums[m] > target) {
            r = m - 1;
        } else if (nums[m] < target) {
            l = m + 1
        } else {
            return m;
        }
    }

    return l;
};

const testCases = [
    {
        nums: [1,3,5,6],
        target: 5,
        expected: 2
    },
    {
        nums: [1,3,5,6],
        target: 2,
        expected: 1
    },
    {
        nums: [1,3,5,6],
        target: 7,
        expected: 4
    },
]

testCases.forEach((testCase, i) => {
    const output = searchInsert(testCase.nums, testCase.target);
    if (JSON.stringify(output) === JSON.stringify(testCase.expected)) {
        console.log(`Test Case ${i}: Passed. Expected: ${JSON.stringify(testCase.expected)}, Output: ${JSON.stringify(output)} `)
    } else {
        console.log(`Test Case ${i}: Failed. Expected: ${JSON.stringify(testCase.expected)}, Output: ${JSON.stringify(output)} `)
    }
});

File: CSharp/35_SearchInsertPosition.cs
namespace LeetCodeSolutions {
    public static class SearchInsertPosition {
        public static int Solution (int[] nums, int target) {
            int m;
            int l = 0;
            int r = nums.Length - 1;

            while (l <= r) {
                m = (l + r) / 2;
                if (nums[m] > target) {
                    r = m - 1;
                } else if (nums[m] < target) {
                    l = m + 1;
                } else {
                    return m;
                }
            }
            return l;
        }
        public static void RunTests() {
            Console.WriteLine("Running Length Of Search Insert Tests...");

            var testCases = new List<(int[], int, int)> {
                (new int[] {1, 3, 5, 6}, 5, 2),
                (new int[] {1, 3, 5, 6}, 2, 1),
                (new int[] {1, 3, 5, 6}, 7, 4),
            };

            foreach (var (nums, target, expected) in testCases) {
                try {
                    var result = Solution(nums, target);
                    Console.WriteLine(
                        $"Input: {string.Join(",", nums)} Target: {target} " +
                        $"Expected: {string.Join(",", expected)} Result: {string.Join(",", result)}"
                    );
                } catch (Exception ex) {
                    Console.WriteLine($"Test Failed: {ex.Message}");
                }
            }
        }
    }
}
